### PR TITLE
BLOCKCHAIN-309 - FE legislation browser  hangs on cancel transaction …

### PR DESCRIPTION
**Pull request by SzymonZiebaNeti**
…- refactor handling errors to same way in sign wallet

I need to refactor all signAndSend functions, to have same error handling, on each one. We have functions which one doing this error handling and showing popup Canceled, and close wallet, if someone cancel sign wallet.

### CODE REVIEW

File: src/api/nodeRpcCall.js [hunk 1]
```diff
@@ -1377,66 +1218,20 @@ const renounceCandidacy = async (walletAddress, userStatus, callback) => {
...
-    }
-  }).catch((error) => {
-    // eslint-disable-next-line no-console
-    console.error(':( elections.submitCandidacy transaction failed', error);
-    callback({ isError: true, details: error.toString() });
-  });
+  return submitExtrinsic(extrinsic, walletAddress, api);
 }
 
 const voteAtMotions = async (walletAddress, proposal, index, vote, callback) => {
```
Comment (kacperzuk-neti) [inline src/api/nodeRpcCall.js:1231]: Callback is now unused. Lots of other functions also still have it.

File: src/redux/sagas/congress.js [hunk 2]
```diff
@@ -68,24 +59,15 @@ function* voteAtMotionsWorker(action) {
   const walletAddress = yield select(
     blockchainSelectors.userWalletAddressSelector,
   );
-
-  const { errorData } = yield cps(
+  yield cps(
```
Comment (kacperzuk-neti) [inline src/redux/sagas/congress.js:62]: This should be changed to `call`. Please double-check all other cases.

**Review by kacperzuk-neti — APPROVED** [review]

File: src/api/nodeRpcCall.js [hunk 3]
```diff
@@ -1,139 +1,148 @@
...
+    blacklistMotion: motions.includes(blacklistMotionHash) ? blacklistMotionHash : null,
+  };
+});
 
-const submitExtrinsic = async (extrinsic, walletAddress) => {
-  const api = await getApi();
-  const { signer } = await web3FromAddress(walletAddress);
-  return await new Promise((resolve, reject) => {
+const submitExtrinsic = async (extrinsic, walletAddress, api, signerFromAdress = true) => {
+  const { signer } = signerFromAdress ? await web3FromAddress(walletAddress) : await web3FromSource('polkadot-js');
```
Comment (DorianSternVukotic) [inline src/api/nodeRpcCall.js:124]: I dont understand what signerfromAddress false is handling.

Comment (kacperzuk-neti) [reply src/api/nodeRpcCall.js:124]: @SzymonZiebaNeti I think we went through it on call and decided that we should always use `await web3FromAddress(walletAddress)` for everything.

**Review by kacperzuk-neti — CHANGES_REQUESTED** [review]
use `await web3FromAddress(walletAddress)` for everything.

**Review by kacperzuk-neti — APPROVED** [review]

File: src/api/nodeRpcCall.js [hunk 4]
```diff
@@ -145,201 +154,157 @@ const getIdentity = async (address) => {
...
-        errorData
-      });
-    }
-  }).catch((error) => {
-    // eslint-disable-next-line no-console
-    console.error(':( transaction failed', error);
-    callback({ isError: true, details: error.toString() });
-  });
-}
+  return submitExtrinsic(call, walletAddress, api, false);
```
Comment (kacperzuk-neti) [inline src/api/nodeRpcCall.js:242]: `submitExtrinsic` no longer takes 4 parameters - only 3.

**Review by kacperzuk-neti — CHANGES_REQUESTED** [review]

File: src/api/nodeRpcCall.js (hunk 3 above)
Comment (DorianSternVukotic) [reply src/api/nodeRpcCall.js]: Could you also explain it to me why this way of doing it was considered ?

Comment (kacperzuk-neti) [reply src/api/nodeRpcCall.js]: The difference is:
* `web3FromAddress(walletAddress)` takes signer for given wallet address from any extension-wallet,
* `web3FromSource('polkadot-js')` takes default wallet from polkadot.js extension.

Hardcoding polkadot-js is obviously wrong, as we want to support other wallets as well.

It wasn't really "considered". @SzymonZiebaNeti just followed how existing code worked during refactor. If you take a look at original code, some functions use `web3FromSource` (https://github.com/liberland/liberland_frontend/blob/20bd59fe38b2faecc9252d61e435069914e57377/src/api/nodeRpcCall.js#L564) and some use `web3FromAddress`.

File: src/api/nodeRpcCall.js [hunk 5]
```diff
@@ -145,201 +154,157 @@ const getIdentity = async (address) => {
...
-        errorData
-      });
-    }
-  }).catch((error) => {
-    // eslint-disable-next-line no-console
-    console.error(':( transaction failed', error);
-    callback({isError: true, details: error.toString()});
-  });
-}
+  return submitExtrinsic(finalCall, walletAddress, api, false);
```
Comment (kacperzuk-neti) [inline src/api/nodeRpcCall.js:292]: Still some 4th parameters left.

**Review by kacperzuk-neti — CHANGES_REQUESTED** [review]

**Review by kacperzuk-neti — APPROVED** [review]

**Pull request merged**